astropy/wcs/utils.py

```diff
@@ -635,3 +635,193 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
+			 b_coeff_names):	 
+		""" Objective function for fitting SIP."""   
+		from ..modeling.models import SIP # here instead of top to avoid circular import
+		
+		crpix = params[0:2]
+		cdx = params[2:6].reshape((2,2))
+		a_params, b_params = params[6:6+len(a_coeff_names)],params[6+len(a_coeff_names):]
+		w_obj.wcs.pc = cdx
+		w_obj.wcs.crpix = crpix
+		x,y = w_obj.wcs_world2pix(lon,lat,1) #'intermediate world coordinates', x & y 
+		x,y = np.dot(w_obj.wcs.pc,(x - w_obj.wcs.crpix[0],y - w_obj.wcs.crpix[1]))
+	
+		a_params, b_params = params[6:6+len(a_coeff_names)], params[6+len(a_coeff_names):]
+		a_coeff, b_coeff = {}, {}
+		for i in range(len(a_coeff_names)):
+			a_coeff['A_' + a_coeff_names[i]] = a_params[i]
+		for i in range(len(b_coeff_names)):
+			b_coeff['B_' + b_coeff_names[i]] = b_params[i]
+			
+		sip = SIP(crpix=crpix, a_order=a_order, b_order=b_order, a_coeff=a_coeff, \
+				  b_coeff=b_coeff)
+		fuv, guv = sip(u,v)
+		xo, yo = np.dot(cdx, np.array([u+fuv-crpix[0], v+guv-crpix[1]]))
+		resids = np.concatenate((x-xo, y-yo))
+		return resids	
+	
+def pixel_sky_to_wcs(xp, yp, coords, projection='TAN', proj_point=None, 
+					 mode='sip', order=None, inwcs=None):	 
+	"""
+	Given a set of matched x,y pixel positions and celestial coordinates, solves for
+	WCS parameters and returns a WCS with these best fit values and other keywords based 
+	on projection type, frame, and units. Optionally, a SIP may be fit. 
+	
+	Along with the matched coordinate pairs, users must provide the projection type 
+	(e.g. 'TAN'), celestial coordinate pair for the projection point (or 'center' to use 
+	the center of input coordinates), mode ('wcs' to fit only linear terms, or 'sip'). 
+	If a coordinate pair is passed to 'proj_point', it is assumed to be in the 
+	same units respectivley as the input (lon, lat) coordinates. Additionaly, if mode is 
+	set to 'sip', the polynomial order must be provided.
+	
+	Optionally, an existing ~astropy.wcs.WCS object with some may be passed in. For 
+	example, this is useful if the user wishes to refit an existing WCS with better
+	astrometry, or are using a projection type with non-standard keywords. If any overlap 
+	between keyword arguments passed to the function and values in the input WCS exist, 
+	the keyword argument will override, including the CD/PC convention. 
+		
+	NAXIS1 and NAXIS2 will set as the span of the input points, unless an input WCS is 
+	provided. 
+	
+	All output will be in degrees. 
+	
+	Parameters
+	----------
+	xp, yp : `numpy.ndarray`
+		X and Y pixel coordinates, respectivley. 
+	coords : `~astropy.coordinates.SkyCoord`
+		Skycoord object.
+	inwcs : None or `~astropy.wcs.WCS`
+		Optional input WCS object. Populated keyword values will be used. Keyword 
+		arguments passed to the function will override any in 'inwcs' if both are present.
+	projection : None or str
+		Three-letter projection code. Defaults to TAN.
+	proj_point: None, str, or tuple of int or float
```

Comment:
I chose to do it this way because celestial_frame_to_wcs accepts a three letter projection code. The projection type can also be determined from ctype if a wcs is input and a string is not.